File: node/src/actor-id.ts
export type ActorIdParts = {
  region: number;
  server: number;
  processType: number;
  processIndex: number;
  actorIndex: number;
};

export type ActorIdCodec = {
  name: string;
  encode(parts: ActorIdParts): bigint;
  decode(value: bigint): ActorIdParts;
};

const REGION_BITS = 10n;
const SERVER_BITS = 12n;
const PROCESS_TYPE_BITS = 6n;
const PROCESS_INDEX_BITS = 10n;
const ACTOR_INDEX_BITS = 26n;

const REGION_MAX = (1n << REGION_BITS) - 1n;
const SERVER_MAX = (1n << SERVER_BITS) - 1n;
const PROCESS_TYPE_MAX = (1n << PROCESS_TYPE_BITS) - 1n;
const PROCESS_INDEX_MAX = (1n << PROCESS_INDEX_BITS) - 1n;
const ACTOR_INDEX_MAX = (1n << ACTOR_INDEX_BITS) - 1n;

const ACTOR_INDEX_SHIFT = 0n;
const PROCESS_INDEX_SHIFT = ACTOR_INDEX_SHIFT + ACTOR_INDEX_BITS;
const PROCESS_TYPE_SHIFT = PROCESS_INDEX_SHIFT + PROCESS_INDEX_BITS;
const SERVER_SHIFT = PROCESS_TYPE_SHIFT + PROCESS_TYPE_BITS;
const REGION_SHIFT = SERVER_SHIFT + SERVER_BITS;

export const defaultActorIdCodec: ActorIdCodec = {
  name: "default",
  encode(parts) {
    const region = BigInt(parts.region);
    const server = BigInt(parts.server);
    const processType = BigInt(parts.processType);
    const processIndex = BigInt(parts.processIndex);
    const actorIndex = BigInt(parts.actorIndex);

    if (region < 0n || region > REGION_MAX ||
        server < 0n || server > SERVER_MAX ||
        processType < 0n || processType > PROCESS_TYPE_MAX ||
        processIndex < 0n || processIndex > PROCESS_INDEX_MAX ||
        actorIndex < 0n || actorIndex > ACTOR_INDEX_MAX) {
      throw new RangeError("ActorId parts out of range");
    }

    return (region << REGION_SHIFT) |
      (server << SERVER_SHIFT) |
      (processType << PROCESS_TYPE_SHIFT) |
      (processIndex << PROCESS_INDEX_SHIFT) |
      (actorIndex << ACTOR_INDEX_SHIFT);
  },
  decode(value) {
    const raw = BigInt(value);
    return {
      region: Number((raw >> REGION_SHIFT) & REGION_MAX),
      server: Number((raw >> SERVER_SHIFT) & SERVER_MAX),
      processType: Number((raw >> PROCESS_TYPE_SHIFT) & PROCESS_TYPE_MAX),
      processIndex: Number((raw >> PROCESS_INDEX_SHIFT) & PROCESS_INDEX_MAX),
      actorIndex: Number((raw >> ACTOR_INDEX_SHIFT) & ACTOR_INDEX_MAX)
    };
  }
};

export function encodeActorId(parts: ActorIdParts, codec: ActorIdCodec = defaultActorIdCodec): bigint {
  return codec.encode(parts);
}

export function decodeActorId(value: bigint, codec: ActorIdCodec = defaultActorIdCodec): ActorIdParts {
  return codec.decode(value);
}

File: go/actor_id.go
package epoch

import "fmt"

type ActorIdParts struct {
	Region       uint16
	Server       uint16
	ProcessType  uint8
	ProcessIndex uint16
	ActorIndex   uint32
}

type ActorIdCodec interface {
	Encode(parts ActorIdParts) (uint64, error)
	Decode(value uint64) ActorIdParts
	Name() string
}

type DefaultActorIdCodec struct{}

const (
	regionBits       = 10
	serverBits       = 12
	processTypeBits  = 6
	processIndexBits = 10
	actorIndexBits   = 26

	regionMax       = (1 << regionBits) - 1
	serverMax       = (1 << serverBits) - 1
	processTypeMax  = (1 << processTypeBits) - 1
	processIndexMax = (1 << processIndexBits) - 1
	actorIndexMax   = (1 << actorIndexBits) - 1

	actorIndexShift   = 0
	processIndexShift = actorIndexShift + actorIndexBits
	processTypeShift  = processIndexShift + processIndexBits
	serverShift       = processTypeShift + processTypeBits
	regionShift       = serverShift + serverBits
)

func (DefaultActorIdCodec) Encode(parts ActorIdParts) (uint64, error) {
	if int(parts.Region) > regionMax ||
		int(parts.Server) > serverMax ||
		int(parts.ProcessType) > processTypeMax ||
		int(parts.ProcessIndex) > processIndexMax ||
		int(parts.ActorIndex) > actorIndexMax {
		return 0, fmt.Errorf("actor id parts out of range")
	}

	value := (uint64(parts.Region) << regionShift) |
		(uint64(parts.Server) << serverShift) |
		(uint64(parts.ProcessType) << processTypeShift) |
		(uint64(parts.ProcessIndex) << processIndexShift) |
		(uint64(parts.ActorIndex) << actorIndexShift)

	return value, nil
}

func (DefaultActorIdCodec) Decode(value uint64) ActorIdParts {
	return ActorIdParts{
		Region:       uint16((value >> regionShift) & regionMax),
		Server:       uint16((value >> serverShift) & serverMax),
		ProcessType:  uint8((value >> processTypeShift) & processTypeMax),
		ProcessIndex: uint16((value >> processIndexShift) & processIndexMax),
		ActorIndex:   uint32((value >> actorIndexShift) & actorIndexMax),
	}
}

func (DefaultActorIdCodec) Name() string {
	return "default"
}

var defaultActorIdCodec ActorIdCodec = DefaultActorIdCodec{}

func EncodeActorId(parts ActorIdParts) (uint64, error) {
	return defaultActorIdCodec.Encode(parts)
}

func DecodeActorId(value uint64) ActorIdParts {
	return defaultActorIdCodec.Decode(value)
}

func DefaultActorIdCodecInstance() ActorIdCodec {
	return defaultActorIdCodec
}
